### ISSUE TITLE
Rancher doesn't realocate container from disconnected host to active one

### ISSUE THREAD

**Issue by Vrtak-CZ** [rancher/rancher#7007]
**Rancher Version:** 1.2.0

**Docker Version:** 1.12.3

**OS and where are the hosts located?:** AWS Autoscaled EC2 Spot instances (RancherOS)

**Setup Details:** single node with external DB (RDS)

**Environment Type:** Cattle

**Steps to Reproduce:** Shut down host (node) and startup new host (node)

**Results:** nothing happend (app / lb etc are not available)

**Expected:** realocate containers from disconected host (node) to active one

<img width="559" alt="screen shot 2016-12-07 at 18 03 25" src="https://cloud.githubusercontent.com/assets/112567/21076253/316ee420-bf27-11e6-8efb-95a76a44e959.png">

**Comment by aemneina** [comment]
@Vrtak-CZ do you have a health check enabled on your service?

**Comment by Vrtak-CZ** [comment]
yes

**Comment by deniseschannon** [comment]
Are there system containers from the infrastructure services on the 2nd host?

**Comment by Vrtak-CZ** [comment]
> Are there system containers from the infrastructure services on the 2nd host?

No. Because they try to start on 1st host with error "Error (instance is not running : Dependencies readiness error)".

If I manually disable 1st host nohting happened, but after I delete 1st host, few minutes later all automatically fix.

**Comment by deniseschannon** [comment]
Are any of the infrastructure services green before you start launching containers? Without those service, health checks will not work and containers will not automatically move to another host.

**Comment by Vrtak-CZ** [comment]
They are allocated on 1st host. So NO they not running. As I said in previous comment. But in rancher UI they show as green and running but on host 1 (so in real they are down).

**Comment by Vrtak-CZ** [comment]
Btw for reproduction just setup rancher server add one service. Add one host kill that host and add another one.

**Comment by Vrtak-CZ** [comment]
Am I doing something wrong? Or is a bug? The reason why we have only one host (node) in this environment is because of staging we don't need HA here.

**Comment by Vrtak-CZ** [comment]
Anyone? cc @deniseschannon

**Comment by deniseschannon** [comment]
If the infrastructure services aren't running, then Rancher won't know about the containers/hosts to be able to move containers as you aren't using the managed overlay network that Rancher supports. 

Can you reproduce tthe fact that you get a host with no infra services?

**Comment by richbos** [comment]
Apologies as didn't see this and have created a duplicate issue. I am experiencing the same behaviour, both in EC2 and in a localised Vagrant/Virtualbox environment. 

https://github.com/rancher/rancher/issues/7420

**Comment by Dmitry1987** [comment]
+1 here, rancher 1.3.1,
just got to the point of testing failover on EC2 spot instances. enabled "rancher integration" in Spotinst (the API key to rancher), spotinst support said they have open issue too, regarding this. their API notified Rancher 2 minutes before instance is going to be terminated. 
Rancher has to use this alert to quickly relocate pods (or at least start the relocation process, start fetching needed images on another host to run containers when the old instance dies).

I'm using kubernetes, latest available. upgraded today.

I think this behavior is same in 1.2.2 and older, probably this is a 'feature request' status? or was it implemented and what we see is a bug/wrong configuration?
I'm on HA 3 masters with external DB.  Prefer to not open another issue if you already discuss it here :) 

BTW it never cleans the "DISCONNECTED" nodes..  any advice on this?  should I script the cleanup through API?
thanks :)

**Comment by jclif** [comment]
I'm also running into something like this, where Hosts removed on SpotInst do not seem to trigger Host deletion in Rancher.

This is the resulting state after 4 spot instances were scaled to 2:

<img width="1439" alt="screen shot 2017-01-17 at 12 23 15 am" src="https://cloud.githubusercontent.com/assets/3463691/22008731/c349510a-dc4b-11e6-8cf4-e8b16dad6f6f.png">

**Comment by Dmitry1987** [comment]
@jclif  did rancher moved your running containers on time? (2 minutes before termination)  you can see exact termination time and compare with container start time ( + - the image pull time of 10-20 seconds for most images?).  I afraid this important functionality stated by Spotinst doesn't work (and it needs fixes from both sides, Rancher and Spotinst, I think).  I'll do more tests, maybe it was a temporary glitch. This functionality is pretty cool, without it we can't keep workloads on spotinst.
I'll try also the K8s integration with the cluster that runs on top of rancher... they have it too, maybe if it works, the underlying rancher behavior will not be that important (if Spotinst will really scale by the size of pod which keeps waiting to be scheduled and has no capacity, and shrink back when less pods are running than needed instances)

**Comment by jclif** [comment]
@Dmitry1987 The containers do not appear to be moved off of the host until after it has disconnected from Rancher. @deniseschannon I will be demoing a Rancher + SpotInst integration to my org next week. Is there any chance of a fix in the short term?

**Comment by Dmitry1987** [comment]
@jclif  i think you can install a script on each spot instance, to look for termination notification at
 http://169.254.169.254/latest/meta-data/spot/termination-time    ([can read here](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-interruptions.html))
when the script sees it got terminate signal, the instance should remove itself from Rancher with API call.  it's a good workaround i think.

**Comment by Dmitry1987** [comment]
@deniseschannon  when we send [DELETE](http://docs.rancher.com/rancher/v1.3/en/api/v2-beta/api-resources/host/)  call, Rancher will immediately start moving pods to other host?  will it first "wait" until the pods are ready on another host and then disconnect host, or first we get disconnected (service downtime on all pods)  and after a wait (k8s detected pods down), pods will be brought up on other hosts?

**Comment by stingerpk** [comment]
I can also confirm this. I am not using Spotinst, using Spot Fleet directly. For reproducing, I terminate an EC2 instance and wait for the spot fleet to create another one. The new instance joins the environment, the infrastructure services are started but the user defined services from the terminated instance are not started. The terminated instance is shown in the disconnected state with user services still showing green. Even the route-53 service has not restarted on the new host.

Rancher single server install with internal DB
Host machines are running RancherOS 0.7.1
Docker version is 1.12.3

![untitled3](https://cloud.githubusercontent.com/assets/9607103/22338090/4fc674b4-e408-11e6-90d2-7789e9cf7257.jpeg)

**Comment by jclif** [comment]
In order to get my Rancher + SpotInst integration working, I had to ensure that I was using an Environment API key. Once I had that in place, the integration worked well.

**Comment by Dmitry1987** [comment]
@jclif  was it Cattle or other env?
I tried env key as you said, but it doesn't work :(  (it didn't connect actually, when I changed the 'account key' to 'env' key)  ,  i'm with Kubernetes though.

**Comment by Dmitry1987** [comment]
oh well, i made it work, not it leaves hosts in "disconnected" yellow state, but migrates the pods!  nice... for the yellow status cleanup (deleting them completely, the old hosts) i'll write periodic script.

**Comment by aemneina** [comment]
related ticket seems to be fixed: https://github.com/rancher/rancher/issues/3239

As thats been verified, I'll close this out. Please file a new ticket if this issue persists on rancher 1.5.9+

**Comment by scomma** [comment]
@Dmitry1987 did you ever end up writing that script? Would you mind sharing it?

**Comment by Dmitry1987** [comment]
@scomma yes, I did the node-cleanup script (removes on cron all the yellow/red ones without asking questions, pretty dangerous!  you can modify it to remove only after N counts that he saw this node 'yellow' [like, put in on 10 minutes cron, and on the 3rd counter that he sees same instance yellow again third time it will delete it. not on the 1st encounter] )

here it is:

```
#!/bin/bash

RANCHER_URL=rancher-dev.yourdomain.com
RANCHER_ACCESS_KEY=xxxxxxxxxxxxxxxxxxxx
RANCHER_SECRET_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

# get project ID easily from the URL when you're in dashboard
PROJECT_ID=XXXXXX

# get all disconnected hosts (YELLOW)
curl -s "https://${RANCHER_URL}/v2-beta/projects/${PROJECT_ID}/hosts" \
         | jq '.data[] |  select(.state | contains("disconnected"))'  | jq -r .id > disconnected_list

# deactivate, before they can be deleted  (make YELLOW to RED)
while read ID ; do
curl  -u "${RANCHER_ACCESS_KEY}:${RANCHER_SECRET_KEY}" \
-X POST -d '{}'  "https://${RANCHER_URL}/v2-beta/projects/${PROJECT_ID}/hosts/${ID}/?action=deactivate"
sleep 1
done < disconnected_list

# allow some time to transition state of all nodes in case rancher overloaded and slow right now
sleep 10

################ HERE I SUGGEST TO MAKE CHANGES LIKE COUNTER ..... 
###########
#############################################################

# get all inactive (RED)
curl -s "https://${RANCHER_URL}/v2-beta/projects/${PROJECT_ID}/hosts" \
         | jq '.data[] |  select(.state | contains("inactive"))'  | jq -r .id > inactive_list

# delete all inactive hosts
while read ID ; do
curl -u "${RANCHER_ACCESS_KEY}:${RANCHER_SECRET_KEY}" \
-X DELETE "https://${RANCHER_URL}/v2-beta/projects/${PROJECT_ID}/hosts/${ID}"
sleep 1
done < inactive_list
```

this script works really well and i had my disconnected spot instances cleaned every time... (i was running this like once in 30 minutes.  the pod migration was taken care of by spotinst API when it was terminating instance, but the yellow nodes remain always in dashboard).

btw we stopped using Rancher after the POC and went with vanilla k8s installed on aws, we had too many little but annoying issues with Rancher in such a short time that we decided it's not worth risking the production environments (Rancher was suddenly moving etcd pods for no reason from node to node, breaking k8s.... and moving or restarting other pods for no visible reason, the UI was crashing and showing 404 or something, etc'). The ideas and approach of Rancher are awesome, but stability not so, the tool seemed too risky for production.  Like, for example: you'd want to decide which db to use for your next project, you try mysql and postgres - and within 1 month you have like 100 glitches and failures in mysql during routine workloads and executing standard queries,  would you proceed with mysql? :)  that was my impression from rancher... i hope others are more lucky and the new versions are stable :)

**Comment by zhangalex** [comment]
Same issue, if there is one node failed then a cluster failed !
